docs/schematypes.md

```diff
@@ -585,13 +585,56 @@ const userSchema = new Schema({
 const User = mongoose.model('User', userSchema);
 ```
 
-To populate every `socialMediaHandles` entry's `oauth` property, you should populate
+To populate every `socialMediaHandles` entry's `oauth` property, you should populatex
 on `socialMediaHandles.$*.oauth`:
 
 ```javascript
 const user = await User.findOne().populate('socialMediaHandles.$*.oauth');
 ```
 
+<h4 id="uuid">UUID</h4>
+
+Mongoose also supports a UUID type that stores UUID v4 instances as [Node.js buffers](https://thecodebarbarian.com/an-overview-of-buffers-in-node-js.html).
+We recommend using [ObjectIds](#objectids) rather than UUIDs for unique document ids in Mongoose, but you may use UUIDs if you need to.
+
+In Node.js, a UUID is represented as an instance of `bson.Binary` type with a [getter](./tutorials/getters-setters.html) that converts the binary to a string when you access it.
+Mongoose stores UUIDs as [binary data with subtype 4 in MongoDB](https://www.mongodb.com/docs/manual/reference/bson-types/#binary-data).
+
+```javascript
+const schema = new mongoose.Schema({
+  docId: {
+    type: mongoose.Schema.Types.UUID // Can also do `type: 'UUID'`
+  }
+});
+const Test = mongoose.model('Test', schema);
+
+const doc = await Test.create({
+  docId: '09190f70-3d30-11e5-8814-0f4df9a59c41'
+});
+
+// 'string'
+console.log(typeof doc.docId);
+// true
+console.log(doc.toObject().docId instanceof mongoose.mongo.BSON.Binary);
+
+const rawDoc = await Test.collection.findOne({ _id: doc._id });
+// true
+console.log(rawDoc.toObject().docId instanceof mongoose.mongo.BSON.Binary);
```

Comment:
Does that mean `docId` is now _the_ unique identifier? Does it have a unique index by default?
Is `_id` still an object id? Can we add `_id: Schema.Types.UUID`?

Also, I don't know how soon or late it is for this, but maybe it'd be a good idea to avoid naming a type "UUID" without the version, something like `UUIDv4` and `UUIDv1` would be easier to maintain and introduce less questions over time.

Reply:
I suppose I can see that being a source of confusion. `docId` is just an arbitrary property name, no particular significance. Maybe instead I should rename to something like `bookSchema` with an `authorId` property.

I'm not familiar enough with UUID to decide whether having v1 or v4 in the name is better or not. I know the MongoDB Node driver just uses UUID, so I think we should leave the name as UUID for consistency, unless we have a strong reason not to.

Reply:
> but maybe it'd be a good idea to avoid naming a type "UUID" without the version, something like UUIDv4 and UUIDv1 would be easier to maintain and introduce less questions over time.
> 
>  I know the MongoDB Node driver just uses UUID, so I think we should leave the name as UUID for consistency, unless we have a strong reason not to.

personally i think we should leave it without a version, because mongoose currently does not validate against a specific version and `Schema.Type.UUID` / `mongo.BSON.Binary`(subtype 4) also does not validate against version and allowing all versions, though it seems like `mongo.BSON.UUID` only supports uuidv4 [see this line](https://github.com/mongodb/js-bson/blob/5d2648ecc03e7ac062978172cbf76b97581cc6f8/src/binary.ts#L457)

Reply:
I was unaware that mongo has a UUID type, so I think going with `UUID` as a name is fine.

I feel like the example can still be confusing. I'd like to propose a full example that can answer a few questions about referencing and uniqueness. Please let me know if the following is valid:

```js
const authorSchema = new Schema({
  _id: Schema.Types.UUID, // Can also do `_id: 'UUID'`
  name: String
});

const Author = mongoose.model('Author', authorSchema);

const bookSchema = new Schema({ 
  authorId: { type: Schema.Types.UUID, ref: 'Author' }
});
const Book = mongoose.model('Book', bookSchema);

const author = new Author({ name: 'Martin Fowler' });
console.log(typeof author._id); // 'string'
console.log(author.toObject()._id instanceof mongoose.mongo.BSON.Binary); // true

const book = new Book({ authorId: '09190f70-3d30-11e5-8814-0f4df9a59c41' });
```